server/src/main/java/org/candlepin/dto/manifest/v1/PoolTranslator.java

```diff
@@ -125,33 +135,73 @@ public PoolDTO populate(ModelTranslator modelTranslator, Pool source, PoolDTO de
+            if (PoolType.STACK_DERIVED.equals(source.getType())) {
+                Set<PoolDTO.ProvidedProductDTO> expectedProvidedProds = new HashSet<>();
+
+                accumulateStackDerivedPoolProvidedProducts(source, expectedProvidedProds);
+
+                if (source.getDerivedProduct() != null) {
+                    for (PoolDTO.ProvidedProductDTO providedProductDTO : expectedProvidedProds) {
+                        dest.addDerivedProvidedProduct(providedProductDTO);
                     }
                 }
-            }
-            else {
-                dest.setProvidedProducts(Collections.emptySet());
-            }
-
-            Set<Product> derivedProducts = source.getDerivedProvidedProducts();
-            if (derivedProducts != null && !derivedProducts.isEmpty()) {
-                for (Product derivedProd : derivedProducts) {
-                    if (derivedProd != null) {
-                        dest.addDerivedProvidedProduct(
-                            new PoolDTO.ProvidedProductDTO(derivedProd.getId(), derivedProd.getName()));
+                else {
+                    for (PoolDTO.ProvidedProductDTO providedProductDTO : expectedProvidedProds) {
+                        dest.addProvidedProduct(providedProductDTO);
                     }
                 }
             }
+        }
+
+        return dest;
+    }
+
+    /**
+     *
+     * @param pool
+     * @return
+     *   accumulated provided products collection from the same stack entitlements
+     */
+    private void accumulateStackDerivedPoolProvidedProducts(Pool pool,
+        Set<PoolDTO.ProvidedProductDTO> expectedProvidedProds) {
+        List<Entitlement> stackedEntitlements = this.entitlementCurator
+            .findByStackId(pool.getSourceStack().getSourceConsumer(), pool.getSourceStackId()).list();
+        for (Entitlement nextStacked : stackedEntitlements) {
+            Pool nextStackedPool = nextStacked.getPool();
+            if (nextStackedPool.getDerivedProduct() == null) {
+                addProvidedProducts(nextStackedPool.getProduct().getProvidedProducts(),
+                    expectedProvidedProds);
+            }
             else {
-                dest.setDerivedProvidedProducts(Collections.emptySet());
+                addProvidedProducts(nextStackedPool.getDerivedProduct().getProvidedProducts(),
+                    expectedProvidedProds);
             }
         }
+    }
 
-        return dest;
+    private void addProvidedProducts(Collection<Product> providedProducts,
+        Set<PoolDTO.ProvidedProductDTO> providedProductDTOs) {
+        if (providedProducts == null || providedProducts.isEmpty()) {
+            return;
+        }
+        for (Product product : providedProducts) {
+            if (product != null) {
+                providedProductDTOs.add(new PoolDTO.ProvidedProductDTO(product.getId(), product.getName()));
+                addProvidedProducts(product.getProvidedProducts(), providedProductDTOs);
```

Comment:
Are these recursive calls what is meant by "n-tier aware"? Are we sure that this is how we're going to be handling this in the future; as a marketing product that has provided products and each provided product has its own provided products etc.? Last I checked we were not going to be doing the N-tier stuff, because we don't yet know how the product management people are going to design that stuff.
@abhiskum @Ceiu

Reply:
"N-tier" the project name and n-tier as the thing I've been saying here are two different things (I think).

I've stated this in another PR already, but for the sake of github comment history, I'll reiterate here:
I posed this question during a post scrum, as I wasn't sure if we should be ignoring the provided products of provided products or not. The discussion ended with "yes, we should handle those" and that's what lead to my (perhaps poorly summarized) comment about n-tier handling.

Whether or not that's something that is or will be used on the business side... who knows. Personally, I feel the effort to write it now and not worry about whether or not they'll use it is so low that I see no reason to skip it, and if it never gets utilized, it's almost free as far as processing time goes. We've probably spent more effort discussing the value than just doing it straight away.

Reply:
While re-reviewing this earlier today, I noticed that the way we handle n-tier here, is that we set all provided products of the hierarchy to the top marketing object.
So if ProductA has ProvidedProductB, and ProvidedProductB itself has ProvidedProductC, then we end up with this DTO structure:
```
ProductA {
    providedProducts: [ ProvidedProductB, ProvidedProductC ]
} 
```
Shouldn't it instead be like this?:
```
ProductA {
    providedProducts: [ 
        ProvidedProductB { 
            providedProducts: [ ProvidedProductC ]
        }
    ]
}
```

Again, sorry if this was already discussed in more detail and I missed it!

Reply:
That should be correct, since this whole remapping is done to simulate the old flat structure we had before.

Reply:
I see, disregard my comment then!